### PR TITLE
Middleware to catch a panic and return a 500 error

### PR DESCRIPTION
**Pull request by bobhageman**
@ivard, please don't hold back;)

### CODE REVIEW

File: server/api.go [hunk 1]
```diff
@@ -554,6 +554,20 @@ func LogMiddleware(typ string, opts LogOptions) func(next http.Handler) http.Han
 	}
 }
 
+// RecoverMiddware is middleware that logs and returns a 500 error if something unhandled (panic) occurs
+func RecoverMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			err := recover()
+			if err != nil {
```
Comment (ivard) [inline server/api.go:562]: A minor thing, recover does not necessarily return an error, so therefore you often see `r := recover()`.

File: server/api.go [hunk 2]
```diff
@@ -554,6 +554,20 @@ func LogMiddleware(typ string, opts LogOptions) func(next http.Handler) http.Han
...
+				LogError(errors.Errorf("Internal server error: %s", err))
```
Comment (ivard) [inline server/api.go:563]: The `%s` only works for strings and that what recover returns does not necessarily have to be a string I think. So, maybe `%v` would be a bit more safe.

**Review by ivard — CHANGES_REQUESTED** [review]

File: server/irmaserver/api.go [hunk 3]
```diff
@@ -170,6 +170,8 @@ func (s *Server) HandlerFunc() http.HandlerFunc {
 	r := chi.NewRouter()
 	s.router = r
 
+	r.Use(server.RecoverMiddleware)
```
[the remark this answers is not in the document]
Comment (bobhageman) [reply server/irmaserver/api.go:173]: Done!

File: server/api.go [hunk 4]
```diff
@@ -554,6 +554,20 @@ func LogMiddleware(typ string, opts LogOptions) func(next http.Handler) http.Han
 	}
 }
 
+// RecoverMiddware is middleware that logs and returns a 500 error if something unhandled (panic) occurs
+func RecoverMiddleware(next http.Handler) http.Handler {
```
Comment (ivard) [inline server/api.go:558]: Small spelling mistake in `RecoverMiddware`.

File: server/keyshare/myirmaserver/server.go [hunk 5]
```diff
@@ -86,6 +86,8 @@ func (s *Server) Handler() http.Handler {
 		router.Use(server.SizeLimitMiddleware)
 		router.Use(server.TimeoutMiddleware(nil, server.WriteTimeout))
 
+		router.Use(server.RecoverMiddleware)
```
Comment (ivard) [inline server/keyshare/myirmaserver/server.go:89]: You've put this everywhere as first middleware to be applied. Maybe good to do that here as well. Then, also a panic in the other middlewares is included.

Comment (bobhageman) [reply server/keyshare/myirmaserver/server.go:89]: Very true, and consistency is key, done

**Review by ivard — APPROVED** [review]

**Pull request merged**